### PR TITLE
No bubbles displayed if only start date exist.

### PR DESCRIPTION
**Pull request by SergeyBondarenko**
Issue #89 
I added the dots by manually manipulating HTML and style. Now it looks exactly like it is in [our docs](https://docs.siren.solutions/kibi/community/4.5.4/#timeline):
screenshot from 2016-12-19 12-16-20

I haven't found a way to do it purely with vis.js. The library doesn't display relations to the X axis when `e.type: 'point'` is used as an event property. In short words, I got the visualization similar to this http://visjs.org/examples/timeline/items/pointItems.html

Other examples: 
http://visjs.org/examples/timeline/items/backgroundAreasWithGroups.html
http://visjs.org/examples/timeline/dataHandling/dataSerialization.html
http://visjs.org/examples/timeline/dataHandling/loadExternalData.html
http://visjs.org/examples/timeline/editing/editingItems.html

Seems like `e.type: point` is designed to not place relations to the X axis.
I can go and fix it in the library.

### CODE REVIEW

**Comment by scampi** [comment]
@SergeyBondarenko please rebase. This PR has changes about the multivalued PR that shouldn't be here.

**Comment by scampi** [comment]
@SergeyBondarenko you can propose a PR to the vis lib if you want ;o)

**Comment by SergeyBondarenko** [comment]
@scampi Yes, I wrote it above. 
Bubbles break your first label emphasize feature. We can make a border around label box to emphasize.

**Comment by scampi** [comment]
> Bubbles break your first label emphasize feature. We can draw a border around label box to emphasize.

good idea.

**Comment by SergeyBondarenko** [comment]
I got "play with CSS" suggestion from one of the vis contributors. As I actually did. https://github.com/almende/vis/issues/2508#issuecomment-269446999

So maybe it is good to take this way.

**Comment by SergeyBondarenko** [comment]
I've fixed the Emphasize feature.
screenshot from 2016-12-28 12-42-42

**Comment by szydan** [comment]
rebased on master

File: public/kibi_timeline.js [hunk 1]
```diff
@@ -235,18 +235,34 @@ define(function (require) {
...
-                  let style = 'background-color: ' + groupColor + '; color: #fff;';
+                  let style = `background-color: ${groupColor}; color: #fff;`;
+                  if (!endFieldValue || startValue === endFieldValue[i]) {
+                    // here the end field value missing but expected
+                    // or start field value === end field value
+                    // force vis box look like vis point
+                    style = `border-style: none; background-color: #fff; color: ${groupColor}; border-color: ${groupColor}`;
+                    const divregex = /(<div.*>)(.*)(<\/div>)/g;
+                    const contentDivParts = divregex.exec(content);
+                    const pointDot = '<div style="position:relative;padding:0;border-width:4px;border-style:solid;' +
```
Comment (scampi) [inline public/kibi_timeline.js]: create a class in the less file with this style.

File: public/kibi_timeline.js [hunk 2]
```diff
@@ -235,18 +235,34 @@ define(function (require) {
...
+                    'border-radius:4px;float:left;margin-top:6px;margin-right:4px;border-color:' + groupColor + '"></div>';
+                    const labelDiv = '<div style="margin-left: 15px">' + contentDivParts[2]  + '</div>';
```
Comment (scampi) [inline public/kibi_timeline.js]: same as above

File: public/kibi_timeline.js [hunk 3]
```diff
@@ -235,18 +235,34 @@ define(function (require) {
...
 
-                  let style = 'background-color: ' + groupColor + '; color: #fff;';
+                  let style = `background-color: ${groupColor}; color: #fff;`;
+                  if (!endFieldValue || startValue === endFieldValue[i]) {
+                    // here the end field value missing but expected
+                    // or start field value === end field value
+                    // force vis box look like vis point
+                    style = `border-style: none; background-color: #fff; color: ${groupColor}; border-color: ${groupColor}`;
+                    const divregex = /(<div.*>)(.*)(<\/div>)/g;
+                    const contentDivParts = divregex.exec(content);
```
Comment (scampi) [inline public/kibi_timeline.js]: Instead of this, you can have an angular template where you would use `ng-class` to set the appropriate class. Then you would compile that template to get the final HTML. See tooltip of the tilemap vis in kibana for some example.

Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: Are you talking about line 250? But this variable is assigned as property to the `e` object below to form the visualization items.

Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: Ahh, ok, you are talking about stuff below

Comment (scampi) [inline public/kibi_timeline.js]: yes ;o) I put the comment here because this way you would have any HTML code in that file and so no need for the regexp

File: public/kibi_timeline.js [hunk 4]
```diff
@@ -237,23 +236,19 @@ define(function (require) {
...
+                    startValue: startValue,
+                    endFieldValue: endFieldValue
                   };
-                  let content = $compile(itemTemplate)($itemscope)[0].innerHTML;
-                  console.log('INNER HTML:');
+
+                  const content = timelineHelper.createItemTemplate(i, itemDict);
                   console.log(content);
 
                   let style = `background-color: ${groupColor}; color: #fff;`;
```
Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: @szydan Seems I can't deprecate this `style` because I need to change the `groupColor` for every group.

File: public/kibi_timeline.js [hunk 5]
```diff
@@ -235,18 +236,35 @@ define(function (require) {
...
+                    endField: params.endField,
+                    labelValue: labelValue,
+                    useHighlight: params.useHighlight,
+                    highlight: timelineHelper.pluckHighlights(hit, highlightTags),
+                    groupColor: groupColor,
+                    startValue: startValue,
+                    endFieldValue: endFieldValue
+                  };
+
+                  const content = timelineHelper.createItemTemplate(i, itemDict);
```
Comment (szydan) [inline public/kibi_timeline.js]: Are you using "i" if not please remove

Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: yes, I do it in the template building method. But I'll look, maybe I can refactor.

Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: I got rid of the i variable in the template method.

File: public/kibi_timeline.js [hunk 6]
```diff
@@ -1,3 +1,4 @@
+/* global angular */
```
Comment (scampi) [inline public/kibi_timeline.js]: why is this needed ?

File: public/lib/helpers/timeline_helper.js [hunk 7]
```diff
@@ -7,6 +7,16 @@ define(function (require) {
     function TimelineHelper() {
     }
 
+    TimelineHelper.prototype.createItemTemplate = function (itemDict) {
+      return '<div title="index: ' + itemDict.indexId + ', startField: ' + itemDict.startField +
```
Comment (scampi) [inline public/lib/helpers/timeline_helper.js]: could you try using lodash template ? https://lodash.com/docs/3.10.1#template
maybe it will improve readability

File: public/kibi_timeline.js (hunk 6 above)
Comment (SergeyBondarenko) [inline public/kibi_timeline.js]: It isn't needed anymore, I'll delete it.

**Comment by SergeyBondarenko** [comment]
jenkins test it

File: public/lib/helpers/timeline_helper.js [hunk 8]
```diff
@@ -7,14 +7,29 @@ define(function (require) {
...
+
     TimelineHelper.prototype.createItemTemplate = function (itemDict) {
-      return '<div title="index: ' + itemDict.indexId + ', startField: ' + itemDict.startField +
-        (itemDict.endField ? ', endField: ' + itemDict.endField : '') + '">' +
-        (!itemDict.endFieldValue || itemDict.startValue === itemDict.endFieldValue
-          ? '<div class="kibi-tl-dot-item" style="border-color:' + itemDict.groupColor + '"></div>' : '') +
-        (!itemDict.endFieldValue || itemDict.startValue === itemDict.endFieldValue
-          ? '<div class="kibi-tl-label-item">' + itemDict.labelValue + '</div>' : itemDict.labelValue) +
-        (itemDict.useHighlight ? '<p class="tiny-txt">' + itemDict.highlight + '</p>' : '') + '</div>';
+      let endfield = '';
```
Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: @scampi how do you find my readability code refactoring?
